Adds basic support for working with system fonts

bushrat011899 opened the pull request:
# Objective

- Fixes #1325
- Contributes to the aside at the end of #16354

## Solution

- Created a new feature, `system_font`, which will instruct `cosmic_text` to begin loading system fonts on startup.
- Added a new system parameter, `FontLibrary`, which allows for convenient searching of all loaded fonts, returning a `Handle<Font>` if you have found a suitable candidate.
- Added a new example, `system_fonts`, which demonstrates searching for and using system fonts.

## Testing

- CI
- `system_fonts` example

---

## Showcase

Users now have access to the `FontLibrary` system parameter, which allows searching for available fonts. This helps with using the new `system_fonts` feature, which will load system fonts at runtime.

_In the below snippets, `fonts` is of type `FontLibrary`_

```rust
// Find a monospaced font
let font: Option<Handle<Font>> = fonts.find(|font| font.monospaced);

// Find a font with a particular name
let font: Option<Handle<Font>> = fonts.find(|font| {
    let family_name = &font.families[0].0;
    family_name.contains("Courier")
});
```

![system_font_demo](https://github.com/user-attachments/assets/42fd4544-d78a-480d-b6f6-32881d18d568)


## Notes

- ~~Loading system fonts increases startup time by a few seconds. I suspect this is platform dependent, and is largely out of our control.~~ The startup cost is now negligible (I cannot measure the difference, much less than a second), since moving system font loading onto a background task. I do believe it's worth keeping this behind a feature gate as it's already organised this way, and there _may_ be a performance cost on some platforms. More options for the user is better IMO.
- If the `system_fonts` feature is enabled on a platform `cosmic_text` does not support, it is simply a no-op, making the feature safe to enable in all cases.
- The `FontLibrary` system param could use some more methods, such as an iterator. I have deliberately only exposed a `find` method since we need to do some work with the `Assets<Font>` to ensure it's available for the `TextPipeline`.
- The `Font` type assumes every font will have binary data associated with it (the `data` field). For system fonts this is not the case. I'm simply inserting no data (empty slice) for these fonts. It might be prudent to update `Font` to be an `enum` instead to make this more clear.
- Font families are still handled the same way Bevy currently does, only selecting index 0. Changing this is out of scope for this PR.

bushrat011899 committed:
Adds basic support for working with system fonts

bushrat011899 committed:
Falling for one of the classic blunders.

bushrat011899 committed:
Falling for the other classic blunder.

bushrat011899 committed:
Clippy

mockersf commented on crates/bevy_text/src/pipeline.rs:
```diff
@@ -32,10 +32,22 @@ pub struct CosmicFontSystem(pub cosmic_text::FontSystem);
+            let mut font_system = cosmic_text::FontSystem::new();
+
+            // This is a no-op on unsupported platforms.
+            font_system.db_mut().load_system_fonts();
```
I suppose this is the one taking time? Is it possible to run this in an asynchronous way, or in another thread?

JeanMertz replied:
Looking at cosmic-text, could we use two [`Database`s](https://docs.rs/fontdb/0.16.2/fontdb/struct.Database.html), one for the system fonts, populated in a different thread, and the main one that loads the application-provided fonts. Then, once the system fonts are loaded, send the database back to the main thread, and merge that one into the main one.

bushrat011899 replied:
Ok what I've done is moved loading the system font onto an `AsyncCompteTaskPool` task. Since I expect this to take multiple frames, and it's entirely in blocking IO, I believe this is the correct pool to use. Once the task resolves, I take these two `Database`s and merge them by simply cloning `FaceInfo`s from one to the other (can use `mem::swap` to copy from the smaller to the larger). I then send a `SystemFontsAvailable` event once this process is completed. It still happens immediately at startup, but it wont prevent other parts of the application from progressing at least.

TurtIeSocks commented:
Awesome! I swapped to this branch locally with my [playground](https://github.com/TurtIeSocks/bevy_simple_i18n/blob/main/examples/changing_locale.rs) example and every locale now displays without the need to load any external font files 🎊 

I suppose this does increase the need for users to have the ability to control `cosmic_text` fallbacks. I should have time later in the week to dig into that and see what's viable.

bushrat011899 commented:
> If it's enabled by default...

I've left it is a feature that's disabled by default. To enable system fonts, you enable the `system_font` feature in Bevy. This feature does have a startup time cost and I completely agree that the functionality should be opt-in, at least while it has a noticeable impact.

TurtIeSocks commented:
I've just noticed #14150, was that effort abandoned?

bushrat011899 commented:
I never saw that! Looks like it wasn't linked to the original issue. I'm happy to incorporate changes from that PR, or close this out in favour of that one, I'll leave that up to the reviewers.

bushrat011899 committed:
Merge remote-tracking branch 'upstream/main' into SystemFonts

bushrat011899 committed:
Fix when upstreaming

bushrat011899 committed:
Merge remote-tracking branch 'upstream/main' into SystemFonts

rparrett commented on crates/bevy_text/src/font_library.rs:
```diff
@@ -0,0 +1,50 @@
+impl FontLibrary<'_> {
+    /// Find a [`Font`] based on the provided criteria.
+    /// You are given access to the font's [`FaceInfo`] to aid with selection.
+    pub fn find(&mut self, mut f: impl FnMut(&FaceInfo) -> bool) -> Option<Handle<Font>> {
```
There may be a subtle performance footgun here for users.

Am I right that calling `find` multiple times for the same font results in multiple asset handles pointing to the same font?

I think this would result in separate font atlases being created and possibly hamper text batching.

Not sure exactly what to do about this, but some ideas:
- Maybe there's a better name for `find` that reflects this behavior
- The behavior should probably be documented
- Maybe the example could be reworked to show something slightly more practical, including caching the font handles?
- Add an ID -> handle mapping? Not sure if that's actually possible

bushrat011899 replied:
Agreed. I might resolve this by adding a `HashMap<FontId, Weak<Font>>` so I can reuse `Handle` if it's already created and still active.

rparrett commented on examples/2d/system_fonts.rs:
```diff
@@ -0,0 +1,74 @@
+            font_size: 50.0,
+            ..default()
+        },
+        TextLayout::new_with_justify(JustifyText::Center),
```
Don't think this is doing anything for an unbounded `Text2d`.

bushrat011899 commented on crates/bevy_text/src/font_library.rs:
```diff
@@ -0,0 +1,50 @@
+///
+/// Note that this is most useful with the `system_font` feature, which exposes
+/// fonts installed on the end-users device. Without this feature, the only fonts
+/// available are ones explicitly added within Bevy anyway.
```
Fair, I could improve the wording here. What I'm trying to say is this allows finding a font and getting a `Handle` to it. But if you have system fonts disabled, it'll only show fonts which you added yourself, and therefore you already have a `Handle` to those fonts.

rparrett replied:
Ahh, okay.

bushrat011899 committed:
Move system font loading to a background task

Also added a mapping between `AssetId<Font>` and `fontdb::ID` to avoid creating new assets and handles when it's not required.

bushrat011899 committed:
Remove redundant `TextLayout`

bushrat011899 committed:
Removed confusing documentation

rparrett commented:
IMO, the current example could be made more practical for users seeking to use this feature (and also for automated testing). It would be nice to show a more common scenario than cycling through every font one at a time, and to demonstrate how to react to the system fonts becoming available.

krunchington commented on examples/2d/system_fonts.rs:
```diff
@@ -0,0 +1,73 @@
+    // Want to make sure we don't show the same font twice
+    mut used: Local<HashSet<Box<str>>>,
+) {
+    if keyboard_input.just_pressed(KeyCode::Space) {
+        let mut name = default();
+
+        // The primary way to find a font is through FontLibrary::find,
+        // which iterates over all loaded fonts, and returns the first that
+        // satisfies the provided predicate.
+        // In this example, we're just looking for any font we haven't already
+        // shown.
+
+        let Some(font) = fonts.find(|font| {
+            let family_name = font.families[0].0.as_str();
+
+            if used.contains(family_name) {
+                return false;
+            }
+
+            used.insert(Box::from(family_name));
+            name = String::from(family_name);
+            true
+        }) else {
+            *query.0 = Text2d::new("No more fonts to show!");
+            return;
+        };
+
+        *query.0 = Text2d::new(name);
+        query.1.font = font;
+    }
```
Nitpick - I think there's an idiomatic way to avoid some returns here.
```suggestion
    if keyboard_input.just_pressed(KeyCode::Space) {
        let mut name = default();
        
        // The primary way to find a font is through FontLibrary::find,
        // which iterates over all loaded fonts, and returns the first that
        // satisfies the provided predicate.
        // In this example, we're just looking for any font we haven't already
        // shown.
        
        let next_font = fonts.find(|font|{
            let family_name = font.families[0].0.as_str();
            if used.contains(family_name) {
                false
            } else {
                used.insert(Box::from(family_name));
                name = String::from(family_name);
                true
            }
        });
        
        match next_font {
            Some(font) => {
                *query.0 = Text2d::new(name);
                query.1.font = font;
            },
            None => {
                *query.0 = Text2d::new("No more fonts to show!");
            },
        }
    }
```

krunchington replied:
Actually I think for the end, `clippy` usually prefers `if let ... else` also but I did this in github comments not my editor XD

krunchington replied:
An example alternate approach that I think ends up a lot more readable, if as in my other comment more info was returned from `.find`
```rs
    if keyboard_input.just_pressed(KeyCode::Space) {
        // The primary way to find a font is through FontLibrary::find,
        // which iterates over all loaded fonts, and returns the first that
        // satisfies the provided predicate.
        // In this example, we're just looking for any font we haven't already
        // shown.
        
        let next_font = fonts.find(|font|{
            let family_name = font.families[0].0.as_str();
            used.contains(family_name)
        });
        
        if let Some((handle, face_info) = next_font {
            let family_name = face_info.families[0].0.as_str();
            *query.0 = Text2d::new(family_name);
            query.1.font = handle;
        } else {
            *query.0 = Text2d::new("No more fonts to show!");
        }
    }
```

krunchington commented on crates/bevy_text/src/font_library.rs:
```diff
@@ -0,0 +1,67 @@
+impl FontLibrary<'_> {
+    /// Find a [`Font`] based on the provided criteria.
+    /// You are given access to the font's [`FaceInfo`] to aid with selection.
+    pub fn find(&mut self, mut f: impl FnMut(&FaceInfo) -> bool) -> Option<Handle<Font>> {
```
It might be useful to return the `FaceInfo` as part of the `Option` also- noticed in the example you had to use a mutable local variable to store data from the FaceInfo.  Maybe something like `Option<(Handle<Font>, &FaceInfo)>`?

krunchington replied:
Also worth noting that this function signature, through the predicate, "leaks" the abstraction of a `cosmic_text::fontdb::FaceInfo`.  That might be intended and fine, but worth pointing out if it is preferable to keep the bevy interface more opaque (which could make it easier to replace `cosmic_text` in the future if necessary)

dsgallups reviewed (approved):
Looks great!
